packages/headless/src/controllers/url-hash-manager/headless-url-hash-manager.ts

```diff
@@ -0,0 +1,196 @@
+   */
+  initialState: UrlHashManagerInitialState;
+}
+
+export interface UrlHashManagerInitialState {
+  /**
+   * The url hash containing the parameters affecting the search response.
+   */
+  urlHash: string;
+}
+
+const urlHashValue = new StringValue({regex: /^#/});
+const initialStateSchema = new Schema<Required<UrlHashManagerInitialState>>({
+  urlHash: urlHashValue,
+});
+
+/**
+ * The `UrlHashManager` controller allows managing an url hash which affect the search response.
+ * */
+export interface UrlHashManager extends Controller {
+  /**
+   * The state relevant to the `UrlHashManager` controller.
+   * */
+  state: UrlHashManagerState;
+  /**
+   * Updates the search parameters from the url hash & launches a search.
+   * @param urlHash The url hash containing the parameters affecting the search.
+   */
+  updateUrlHash(urlHash: string): void;
+}
+
+export interface UrlHashManagerState {
+  /**
+   * The url hash containing the parameters affecting the search response.
+   */
+  urlHash: string;
+}
+
+/**
+ * Creates a `UrlHashManager` controller instance.
+ *
+ * @param engine - The headless engine.
+ * @param props - The configurable `UrlHashManager` properties.
+ * @returns A `UrlHashManager` controller instance.
+ */
+export function buildUrlHashManager(
+  engine: Engine<Partial<SearchParametersState> & ConfigurationSection>,
+  props: UrlHashManagerProps
+): UrlHashManager {
+  const {dispatch} = engine;
+  const controller = buildController(engine);
+
+  validateInitialState(
+    engine,
+    initialStateSchema,
+    props.initialState,
+    'buildUrlHashManager'
+  );
+
+  const parameters = {
+    ...getInitialSearchParameterState(engine),
+    ...getSearchParameterStateFromUrlHash(props.initialState.urlHash),
+  };
+  const searchParameterManager = buildSearchParameterManager(engine, {
+    initialState: {
+      parameters,
+    },
+  });
+
+  return {
+    ...controller,
+
+    get state() {
+      return {
+        urlHash: buildSearchParameterSerializer().serialize(
+          searchParameterManager.state.parameters
+        ),
+      };
+    },
+
+    updateUrlHash(urlHash: string) {
+      const validation = urlHashValue.validate(urlHash);
+      if (validation) {
+        engine.logger.error(validation, urlHash, 'updateUrlHash');
+        return;
+      }
+
+      const initialParameters = getInitialSearchParameterState(engine);
+      const previousParameters = {
+        ...initialParameters,
+        ...searchParameterManager.state.parameters,
+      };
+      const newParameters = {
+        ...initialParameters,
+        ...getSearchParameterStateFromUrlHash(urlHash),
+      };
+      dispatch(restoreSearchParameters(newParameters));
+      dispatch(
+        executeSearch(
+          getRelevantAnalyticsAction(previousParameters, newParameters)
+        )
+      );
+    },
+  };
+}
+
+function getSearchParameterStateFromUrlHash(urlHash: string) {
+  const urlWithoutHash = urlHash.slice(1);
```

Comment:
That's a flaky assumption. 

Could use some Regex magic to try to determine if somebody passes in a string that start with `#`, or `?`, or no such leading character.